<a href="https://colab.research.google.com/github/Mirko6/part-II-project/blob/main/pyKT_evals/pyKT_isaac_all_models.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

## pyKT install

In [ ]:
%pip install -U pykt-toolkit

Looking in indexes: https://pypi.org/simple, https://us-python.pkg.dev/colab-wheels/public/simple/
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 114.3/114.3 kB 6.2 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 2.0/2.0 MB 15.0 MB/s eta 0:00:00
  Preparing metadata (setup.py) ... done
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 201.7/201.7 kB 11.6 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 184.3/184.3 kB 14.6 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 62.7/62.7 kB 3.5 MB/s eta 0:00:00
  Created wheel for pathtools: filename=pathtools-0.1.2-py3-none-any.whl size=8807 sha256=71274e0d53856366d30b5a44eaeefd04d52861a340df25d77be953bacca3ebca
  Stored in directory: /root/.cache/pip/wheels/e7/f3/22/152153d6eb222ee7a56ff8617d80ee5207207a8c00a7aab794
Successfully built pathtools


In [ ]:
!git clone https://github.com/pykt-team/pykt-toolkit.git

Cloning into 'pykt-toolkit'...
remote: Enumerating objects: 6327, done.
remote: Counting objects: 100% (737/737), done.
remote: Compressing objects: 100% (310/310), done.
remote: Total 6327 (delta 530), reused 595 (delta 426), pack-reused 5590
Receiving objects: 100% (6327/6327), 33.81 MiB | 16.72 MiB/s, done.
Resolving deltas: 100% (4911/4911), done.


## assist 2009
just for creating the file path data/assist2009/skill_....csv

I uploaded the isaac csv manually

In [ ]:
#assistment 2009 - download
assistment_2009_download_id = '1NNXHFRxcArrU0ZJSb9BIL56vmUt5FhlE'
# google link 'https://drive.google.com/u/0/uc?id=1NNXHFRxcArrU0ZJSb9BIL56vmUt5FhlE&export=download'
url = 'https://drive.google.com/uc?id=' + assistment_2009_download_id
import gdown, pandas as pd
!mkdir pykt-toolkit/data/assist2009
output = 'pykt-toolkit/data/assist2009/skill_builder_data_corrected_collapsed.csv'
gdown.download(url, output, quiet=False)

mkdir: cannot create directory ‘pykt-toolkit/data/assist2009’: File exists


Downloading...
From: https://drive.google.com/uc?id=1NNXHFRxcArrU0ZJSb9BIL56vmUt5FhlE
To: /content/pykt-toolkit/data/assist2009/skill_builder_data_corrected_collapsed.csv
100%|██████████| 64.4M/64.4M [00:00<00:00, 271MB/s]


'pykt-toolkit/data/assist2009/skill_builder_data_corrected_collapsed.csv'

In [ ]:
#assistment 2009 - open using windows-1252 encoding and save to appropriate place

df = pd.read_csv(
    "pykt-toolkit/data/assist2009/skill_builder_data_corrected_collapsed.csv",
    encoding='windows-1252'
  )
df_sample = df.head(500) # to have a faster execution, comment out this line to use whole dataset
df_sample.to_csv("pykt-toolkit/data/assist2009/skill_builder_data_corrected_collapsed.csv")
df_sample.head()

<ipython-input-9-45fbecc8863a>:3: DtypeWarning: Columns (17,18) have mixed types. Specify dtype option on import or set low_memory=False.
  df = pd.read_csv(


Unnamed: 0  order_id  assignment_id  user_id  assistment_id  problem_id  \
0           1  33022537         277618    64525          33139       51424   
1           2  33022709         277618    64525          33150       51435   
2           3  35450204         220674    70363          33159       51444   
3           4  35450295         220674    70363          33110       51395   
4           5  35450311         220674    70363          33196       51481   

   original  correct  attempt_count  ms_first_response  ... hint_count  \
0         1        1              1              32454  ...          0   
1         1        1              1               4922  ...          0   
2         1        0              2              25390  ...          0   
3         1        1              1               4859  ...          0   
4         1        0             14              19813  ...          3   

  hint_total  overlap_time  template_id  answer_id answer_text  first_action  \
0          3         32454        30799        NaN          26             0   
1          3          4922        30799        NaN          55             0   
2          3         42000        30799        NaN          88             0   
3          3          4859        30059        NaN          41             0   
4          4        124564        30060        NaN          65             0   

  bottom_hint opportunity  opportunity_original  
0         NaN           1                   1.0  
1         NaN           2                   2.0  
2         NaN           1                   1.0  
3         NaN           2                   2.0  
4         0.0           3                   3.0  

[5 rows x 31 columns]

In [ ]:
print(f'number of interactions: {len(df)}')
print(f"number of questions: {len(df['problem_id'].unique())}")
print(f"number of students: {len(df['user_id'].unique())}")

number of interactions: 346860
number of questions: 26688
number of students: 4217


In [ ]:
#df.to_csv("pykt-toolkit/data/assist2009/skill_builder_data_corrected_collapsed.csv")

Make sure, I use the same data as pyKT:


https://pykt-toolkit.readthedocs.io/en/latest/datasets.html
This dataset is made up of math exercises, collected from the free online tutoring ASSISTments platform in the school year 2009-2010. The dataset consists of **346,860 interactions, 4,217 students, and 26,688 questions** and is widely used and has been the standard benchmark for KT methods over the last decade.

## Preprocess

In [ ]:
!cd pykt-toolkit/examples/ ; python data_preprocess.py --dataset_name=assist2009

Namespace(dataset_name='assist2009', file_path='../data/peiyou/grade3_students_b_200.csv', min_seq_len=3, maxlen=200, kfold=5)
Start preprocessing data: assist2009
original interaction num: 480534, user num: 46612, question num: 536, concept num: 94, avg(ins) per s: 10.3092, avg(c) per q: 2.4272, na: 0
after drop interaction num: 480534, user num: 46612, question num: 536, concept num: 94, avg(ins) per s: 10.3092, avg(c) per q: 2.4272, na: 0
/usr/local/lib/python3.10/dist-packages/pykt/preprocess/assist2009_preprocess.py:25: FutureWarning: In a future version of pandas, a length 1 tuple will be returned when iterating over a groupby with a grouper equal to a list of length 1. Don't supply a list with a single grouper to avoid this warning.
  for ui in ui_df:
480534,46612,536,94,10.3092,2.4272,0
480534,46612,536,94,10.3092,2.4272,0
--------------------------------------------------
delete bad stu num of len: 16624, delete interactions: 22934, of r: 0, good num: 457600
original total int

# train and evals

## DKT

In [ ]:
!cd pykt-toolkit/examples/ ; python3 wandb_dkt_train.py --dataset_name=assist2009 --use_wandb=0 > dkt_train.txt

In [ ]:
!cd pykt-toolkit/examples/ ; python3 wandb_predict.py --use_wandb=0 --save_dir='saved_model/assist2009_dkt_qid_saved_model_42_0_0.2_200_0.001_0_1'

Namespace(bz=256, save_dir='saved_model/assist2009_dkt_qid_saved_model_42_0_0.2_200_0.001_0_1', fusion_type='early_fusion,late_fusion', use_wandb=0)
model_name is dkt
Read data from processed file: ../data/assist2009/test_sequences.csv_-1.pkl
file path: ../data/assist2009/test_sequences.csv, qlen: 6235, clen: 6235, rlen: 6235
Start preprocessing ../data/assist2009/test_window_sequences.csv fold: _-1...
interaction_num: 235588
file path: ../data/assist2009/test_window_sequences.csv, qlen: 33311, clen: 33311, rlen: 33311
Start preprocessing ../data/assist2009/test_question_sequences.csv fold: _-1...
interaction_num: 220892
file path: ../data/assist2009/test_question_sequences.csv, qlen: 198021, clen: 198021, rlen: 198021
Start preprocessing ../data/assist2009/test_question_window_sequences.csv fold: _-1...
interaction_num: 221317
file path: ../data/assist2009/test_question_window_sequences.csv, qlen: 201109, clen: 201109, rlen: 201109
has test_question_file!
Start predicting model: dkt, 

## DKT+

In [ ]:
!cd pykt-toolkit/examples/ ; python3 wandb_dkt_plus_train.py --dataset_name=assist2009 --use_wandb=0 > dkt_plus_train.txt

In [ ]:
!cd pykt-toolkit/examples/ ; python3 wandb_predict.py --use_wandb=0 --save_dir='saved_model/assist2009_dkt+_qid_saved_model_42_0_0.2_200_0.001_0.01_0.003_3.0_0_1'

Namespace(bz=256, save_dir='saved_model/assist2009_dkt+_qid_saved_model_42_0_0.2_200_0.001_0.01_0.003_3.0_0_1', fusion_type='early_fusion,late_fusion', use_wandb=0)
model_name is dkt+
Read data from processed file: ../data/assist2009/test_sequences.csv_-1.pkl
file path: ../data/assist2009/test_sequences.csv, qlen: 6235, clen: 6235, rlen: 6235
Read data from processed file: ../data/assist2009/test_window_sequences.csv_-1.pkl
file path: ../data/assist2009/test_window_sequences.csv, qlen: 33311, clen: 33311, rlen: 33311
Read data from processed file: ../data/assist2009/test_question_sequences.csv_-1_qtest.pkl
file path: ../data/assist2009/test_question_sequences.csv, qlen: 198021, clen: 198021, rlen: 198021
Read data from processed file: ../data/assist2009/test_question_window_sequences.csv_-1_qtest.pkl
file path: ../data/assist2009/test_question_window_sequences.csv, qlen: 201109, clen: 201109, rlen: 201109
has test_question_file!
Start predicting model: dkt+, embtype: qid, save_dir: sav

## DKVMN

In [ ]:
!cd pykt-toolkit/examples/ ; python3 wandb_dkvmn_train.py --dataset_name=assist2009 --use_wandb=0 > dkvmn_train.txt

In [ ]:
!cd pykt-toolkit/examples/ ; python3 wandb_predict.py --use_wandb=0 --save_dir='saved_model/assist2009_dkvmn_qid_saved_model_42_0_0.2_200_0.001_50_0_1'

Namespace(bz=256, save_dir='saved_model/assist2009_dkvmn_qid_saved_model_42_0_0.2_200_0.001_50_0_1', fusion_type='early_fusion,late_fusion', use_wandb=0)
model_name is dkvmn
Read data from processed file: ../data/assist2009/test_sequences.csv_-1.pkl
file path: ../data/assist2009/test_sequences.csv, qlen: 6235, clen: 6235, rlen: 6235
Read data from processed file: ../data/assist2009/test_window_sequences.csv_-1.pkl
file path: ../data/assist2009/test_window_sequences.csv, qlen: 33311, clen: 33311, rlen: 33311
Read data from processed file: ../data/assist2009/test_question_sequences.csv_-1_qtest.pkl
file path: ../data/assist2009/test_question_sequences.csv, qlen: 198021, clen: 198021, rlen: 198021
Read data from processed file: ../data/assist2009/test_question_window_sequences.csv_-1_qtest.pkl
file path: ../data/assist2009/test_question_window_sequences.csv, qlen: 201109, clen: 201109, rlen: 201109
has test_question_file!
Start predicting model: dkvmn, embtype: qid, save_dir: saved_model/

## SAINT

In [ ]:
!cd pykt-toolkit/examples/ ; python3 wandb_saint_train.py --dataset_name=assist2009 --use_wandb=0 > saint_train.txt

In [ ]:
!cd pykt-toolkit/examples/ ; python3 wandb_predict.py --use_wandb=0 --save_dir='saved_model/assist2009_saint_qid_saved_model_42_0_0.2_256_0.001_8_4_0_1'

Namespace(bz=256, save_dir='saved_model/assist2009_saint_qid_saved_model_42_0_0.2_256_0.001_8_4_0_1', fusion_type='early_fusion,late_fusion', use_wandb=0)
model_name is saint
Read data from processed file: ../data/assist2009/test_sequences.csv_-1.pkl
file path: ../data/assist2009/test_sequences.csv, qlen: 6235, clen: 6235, rlen: 6235
Read data from processed file: ../data/assist2009/test_window_sequences.csv_-1.pkl
file path: ../data/assist2009/test_window_sequences.csv, qlen: 33311, clen: 33311, rlen: 33311
Read data from processed file: ../data/assist2009/test_question_sequences.csv_-1_qtest.pkl
file path: ../data/assist2009/test_question_sequences.csv, qlen: 198021, clen: 198021, rlen: 198021
Read data from processed file: ../data/assist2009/test_question_window_sequences.csv_-1_qtest.pkl
file path: ../data/assist2009/test_question_window_sequences.csv, qlen: 201109, clen: 201109, rlen: 201109
has test_question_file!
Start predicting model: saint, embtype: qid, save_dir: saved_model

## SAKT

In [ ]:
!cd pykt-toolkit/examples/ ; python3 wandb_sakt_train.py --dataset_name=assist2009 --use_wandb=0 > sakt_train.txt

In [ ]:
!cd pykt-toolkit/examples/ ; python3 wandb_predict.py --use_wandb=0 --save_dir='saved_model/assist2009_sakt_qid_saved_model_42_0_0.2_256_0.001_8_1_0_1'

Namespace(bz=256, save_dir='saved_model/assist2009_sakt_qid_saved_model_42_0_0.2_256_0.001_8_1_0_1', fusion_type='early_fusion,late_fusion', use_wandb=0)
model_name is sakt
Read data from processed file: ../data/assist2009/test_sequences.csv_-1.pkl
file path: ../data/assist2009/test_sequences.csv, qlen: 6235, clen: 6235, rlen: 6235
Read data from processed file: ../data/assist2009/test_window_sequences.csv_-1.pkl
file path: ../data/assist2009/test_window_sequences.csv, qlen: 33311, clen: 33311, rlen: 33311
Read data from processed file: ../data/assist2009/test_question_sequences.csv_-1_qtest.pkl
file path: ../data/assist2009/test_question_sequences.csv, qlen: 198021, clen: 198021, rlen: 198021
Read data from processed file: ../data/assist2009/test_question_window_sequences.csv_-1_qtest.pkl
file path: ../data/assist2009/test_question_window_sequences.csv, qlen: 201109, clen: 201109, rlen: 201109
has test_question_file!
Start predicting model: sakt, embtype: qid, save_dir: saved_model/as